Add username unavailability indicator and validation to Edit Profile …

4555jan opened the pull request:
## Description
This PR adds the username unavailability indicator feature to the Edit Profile Screen, matching the functionality already present in the Onboarding Screen. Previously, the Edit Profile Screen only showed when a username was available but provided no feedback when a username was unavailable or invalid.

**Changes made:**
- Added "Checking..." indicator while validating username
- Added red snackbar notification when username is unavailable
- Added client-side validation for username format (letters, numbers, dots, hyphens, underscores only)
- Added 36 character limit to match Appwrite document ID constraints
- Added `usernameAvailableChecking` observable to track validation state

Fixes #440 





https://github.com/user-attachments/assets/f7a06d07-0c38-44c7-92e3-c97318d5dd80









## Type of change
- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
## How Has This Been Tested?
**Manual Testing:**
1. Navigate to Edit Profile Screen
2. Attempt to change username with various inputs:
   - Valid username (6+ chars, valid format) - Shows green checkmark 
   - Invalid characters (spaces, special chars) - Shows error message 
   - Too long (>36 chars) - Shows error message 
   - Too short (<6 chars) - Shows error message 
   - Existing username - Shows "Username unavailable" snackbar 


**Test Configuration:**
- Device: android emulator
- 
## Checklist:
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules
- [x] I have checked my code and corrected any misspellings

## Maintainer Checklist
- [x] closes #440 
- [x] Tag the PR with the appropriate labels

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Real-time username validation and availability check with debounce during profile editing
  * Live visual feedback: checking spinner, verified or unavailable icons, and snackbar for unavailable names
  * Save button now respects validation and availability; max username length enforced (36 chars)

* **Localization**
  * Added/updated messages for invalid format, already-taken username, and deleted/failed-delete notices in multiple languages

* **Tests**
  * Coverage added/updated for username availability scenarios

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

4555jan committed:
Add username unavailability indicator and validation to Edit Profile Screen

4555jan commented:
sorry i will sure make sure of that @M4dhav in the future i think i made a minor/major  mistake with my commit in the files i will fix it soon are these changes that i made is mergeable tho???

M4dhav commented:
Currently reviewing, will let you know

M4dhav reviewed (commented):
Please share a video of the functionality

4555jan committed:
fix:mistakenly made another variable for username checking in controller the variable was 'username Available'

M4dhav commented:
Okay, design looks fine. One small change though, we have a minimum length of username, it is 6 or 7, you need to check for the exact value. Below that length the check should not trigger. Also, how does the UI react when a username is taken?

4555jan commented:
@M4dhav i uploaded another video regarding how the ui will change if the username is taken

4555jan committed:
changed the minimum length to 7

4555jan committed:
kept the milliseconds to 800 like before

4555jan committed:
solved a localization error

M4dhav commented on lib/views/screens/edit_profile_screen.dart:
```diff
@@ -131,19 +140,55 @@ class EditProfileScreen extends StatelessWidget {
+                                AppLocalizations.of(
+                                  context,
+                                )!.usernameUnavailable,
+                                "Username can only contain letters, numbers, dots, hyphens, and underscores (max 36 characters)",
```
Add localizations

M4dhav commented on lib/views/screens/edit_profile_screen.dart:
```diff
@@ -131,19 +140,55 @@ class EditProfileScreen extends StatelessWidget {
+                            }
+
+                            // Temporarily show checking state
+                            controller.usernameAvailable.value = true;
```
This is not a checking state, this will show the user a green tick even while checking

M4dhav reviewed (changes requested):
Please run `flutter gen-l10n` with each PR to check untranslated.txt into source

4555jan committed:
Merge branch 'dev' into feat-username-validation

4555jan committed:
Fix username validation and add proper checking state

M4dhav commented:
Yes that is fine. Could you share an updated video of the functionalituy as well

M4dhav commented on lib/views/screens/edit_profile_screen.dart:
If you are no longer opening snackbars for validation then this is unnecessary

4555jan replied:
We are still showing snackbars for validation errors (invalid format and username taken), so Get.closeCurrentSnackbar() is needed isnt

M4dhav commented on lib/views/screens/edit_profile_screen.dart:
```diff
@@ -184,14 +179,17 @@ class EditProfileScreen extends StatelessWidget {
                             });
                           } else {
                             controller.usernameAvailable.value = false;
+                            controller.usernameChecking.value = false;
```
These are unnecessary as you are already settings the values above

4555jan committed:
separate username validation error messages and removed unnecessary line

4555jan committed:
removed an unnecessary string

M4dhav reviewed (changes requested):
I think we should disable the save changes button when the username is not valid, as the autovalidator is already showing the necessary feedback inline

4555jan committed:
fix: added description for last four strings

4555jan committed:
fix: disabled the save changes button when the username is invalid

M4dhav commented on lib/controllers/edit_profile_controller.dart:
Wrap this in the try catch below as well

4555jan committed:
fix: Remove username availability check and added try catch block

M4dhav commented on lib/controllers/edit_profile_controller.dart:
```diff
@@ -259,52 +259,35 @@ class EditProfileController extends GetxController {
-          data: {'email': authStateController.email},
-        );
-
         try {
```
Why 2 try catch blocks? same block can handle all conditions. Also, as you are rethrowing errors,ensure that errors are captured in the view and appropriately displayed, for example, via snackbar

4555jan replied:
but we are displaying the errors in snakebar already isnt i am sorry ig rethrowing the errors is the problem is it? @M4dhav

M4dhav commented:
Just the ones you added in this PR

4555jan committed:
Merge branch 'dev' into feat-username-validation

4555jan committed:
Fix single try and catch block

M4dhav reviewed (commented):
Please share a video of the functionality as it looks right now in a new comment

M4dhav commented on lib/views/screens/edit_profile_screen.dart:
```diff
@@ -131,22 +141,65 @@ class EditProfileScreen extends StatelessWidget {
+                                  snackbarDuration: 1,
                                 );
+                              }
+                            });
```
All this logic can also be moved to the validator

4555jan committed:
removed the snakebars and moved to the validator

4555jan commented:
@M4dhav i am sorry  i can remove that snake bar as well? even i got that doubt after commit

4555jan commented:
https://github.com/user-attachments/assets/d715b895-de4d-4051-9ccd-0bcfa5a33097

is it correct? i was having some branch issues in the main repo earlier with  upstream and all so i tested it like this thats why there is no M beside the files

4555jan committed:
Fix fixed save profile button and icon when users

M4dhav commented on lib/l10n/app_en.arb:
Please add translations for these strings for your native languages

4555jan committed:
fix: reduced unnecessary conditions

4555jan committed:
fix: translations

4555jan commented:
@M4dhav do i need to sent the video ? it works similar way like the old one i sented ...

M4dhav commented:
> @M4dhav do i need to sent the video ? it works similar way like the old one i sented ...

Yes please do

4555jan commented:
https://github.com/user-attachments/assets/e13af694-730f-44fe-81f8-450bfff85124


is it correct? now ? @M4dhav

M4dhav commented:
When the debouncer runs and username is found to be unavailable, shouldn't the text field have a Red cross icon at the end? Also shorten the validator text as now it is only in the validator and we have less space there

4555jan commented:
<img width="1920" height="1080" alt="image" src="https://github.com/user-attachments/assets/4a28b13c-00c7-47cc-bf6e-4a5bf657178a" />
<img width="1920" height="1080" alt="image" src="https://github.com/user-attachments/assets/ff0ad76f-af34-46ed-8764-3d3866b46278" />




something like this?
@M4dhav so red cross should show when dealing with user unavailable ? just for this error? if thats the case than i will put a condition in the suffixicon for that is this a good  solution ? or we just omit the snakebar for this one as well ?




<img width="743" height="393" alt="image" src="https://github.com/user-attachments/assets/a70f7b9b-49b9-442a-9919-1e8c1fb5d447" />



this is how it can be implemented with a circular indicator

M4dhav commented:
Was there nothing there before? How does the other place with the check handle it?

4555jan commented:
> Was there nothing there before? How does the other place with the check handle it?

please elaborate.........i didnt understood ...... what is actually breaking in the feature  ? if you mean by the suffixicon than it was just green before just the way it is in the dev repo before i made all this change in this feature @M4dhav 

<img width="498" height="163" alt="image" src="https://github.com/user-attachments/assets/4eeb46bd-1e9f-4b54-b3a5-91864cc2b5ff" /> it was just showing green in here as well

M4dhav commented:
That is surprising.  Red cross is fine, please push the code and tag me here.

4555jan committed:
fix: username validation showing red X if username is avilable

4555jan commented:
https://github.com/user-attachments/assets/c532a78f-2dc5-488a-a3b0-b57bfdd57627

is it working now as u expected?.......correct?  @M4dhav

M4dhav commented on lib/controllers/edit_profile_controller.dart:
```diff
@@ -26,7 +26,7 @@ class EditProfileController extends GetxController {
 
   RxBool isLoading = false.obs;
-  Rx<bool> usernameAvailable = false.obs;
+  Rx<bool> usernameAvailable = true.obs;
```
why initialize it to true?

M4dhav commented on lib/views/screens/edit_profile_screen.dart:
```diff
@@ -181,19 +189,34 @@ class EditProfileScreen extends StatelessWidget {
+                                            .length >=
+                                        7 &&
+                                    RegExp(r'^[a-zA-Z0-9._-]+$').hasMatch(
+                                      controller.usernameController.text.trim(),
```
Why are these conditions needed?

4555jan commented:
Because of implementing the cross feature like if we just made changes to suffixicon it was showing cross even when the user owns it's own username @M4dhav

M4dhav commented:
> Because of implementing the cross feature like if we just made changes to suffixicon it was showing cross even when the user owns it's own username @M4dhav

This is why we are initializing it to true right? In that case it is fine. Please answer the other question, as I believe the extra conditions are unnecessary.

Also, make sure to modify the submit button to not run the username updation logic if the user hasn't actually changed the username

4555jan committed:
fix: make sure that save button doesnt call api on current username and red x works correctly

4555jan committed:
fix: checks were failing for usernameavilable so had to edit testfile for specific username check when user own's its username

4555jan committed:
fix: checks were failing

4555jan commented:
https://github.com/user-attachments/assets/2c2bfa9a-3e93-4b00-8299-1d38eccf0370



in edit controller  function isUsernameAvailable added check for if the is entering its old username if yes return true and in onit method made usernameAvailable=true since a user starts with their own username  
had to edit test file for isusername available

M4dhav commented on lib/views/screens/edit_profile_screen.dart:
```diff
@@ -189,34 +181,20 @@ class EditProfileScreen extends StatelessWidget {
-                                  color: Colors.green,
-                                )
-                              : (controller.usernameController.text
-                                            .trim()
```
Circular Progress Indicator was a good idea, add that back in, but no need for the last condition that was checking for validation

M4dhav commented on lib/views/screens/edit_profile_screen.dart:
```diff
@@ -145,23 +145,15 @@ class EditProfileScreen extends StatelessWidget {
-
                           if (trimmedValue.length < 7) {
                             controller.usernameAvailable.value = false;
-                            controller.usernameChecking.value = false;
```
Why deleted

4555jan replied:
earlier errors were comming because of that...... but now  was some other issue not because of that line i realized  i fixed it and added that check again like before

M4dhav commented on lib/views/screens/edit_profile_screen.dart:
```diff
@@ -145,23 +145,15 @@ class EditProfileScreen extends StatelessWidget {
                             r'^[a-zA-Z0-9._-]+$',
                           ).hasMatch(trimmedValue)) {
                             controller.usernameAvailable.value = false;
-                            controller.usernameChecking.value = false;
```
Why deleted

M4dhav commented on test/controllers/edit_profile_controller_test.dart:
```diff
@@ -67,8 +67,9 @@ void main() {
     expect(editProfileController.usernameController.text, 'testuser');
   });
   test('check isUsernameAvailable', () async {
+    // user's own username should return true
     final result = await editProfileController.isUsernameAvailable('testuser');
-    expect(result, false);
+    expect(result, true);
   });
```
Add another test here to go with this change, that verifies the function for both cases

4555jan committed:
fix: added test for both functions and circular controller added back

4555jan committed:
fix: added mock for another user

4555jan commented:
https://github.com/user-attachments/assets/eaf8ea00-e505-42cc-9282-007266c5c439

added circular progress indicator   added those deleted line back  also added mock for checking another user because before the checks were failing

M4dhav commented:
@coderabbitai review

M4dhav commented:
Hey @4555jan I think coderabbit may be on to something here, if username change fails we should rethrow or set the showSuccess var to false

M4dhav commented:
just change the var to false

4555jan committed:
fix: rethrow showSuccessSnackbar

4555jan commented:
<img width="502" height="550" alt="image" src="https://github.com/user-attachments/assets/e975dff5-5062-434a-a445-ec3cfd38f392" />
like this right  @M4dhav

M4dhav commented:
Great work on this PR, and sticking with it for all these months!
Merging
